Issue/12342/js test driver removal

KarlDeux opened the pull request:
Fixes #12342 

Refactoring and creation of Jasmine Unit testing.
Removed JSTestDriver.

<!--- Provide a general summary of the Pull Request in the Title above -->

### Description
<!--- Provide a description of the changes proposed in the pull request -->
Currently Magento is using 2 different sets of tools for the JavaScript Testing:

JSTestDriver, which is considered legacy and is not supported anymore by the core team.
Jasmine, which was introduced to replace JSTestDriver. All new JS tests are implemented using it and executed in multiple CI environments, including public Travis CI and Magento's in-house CICD infrastructure.
To remove legacy framework it is required to reimplement remaining tests using Jasmine, and completely remove JSTestDriver support afterwards.

### Fixed Issues (if relevant)
<!--- Provide a list of fixed issues in the format magento/magento2#<issue_number>, if relevant  -->
1. magento/magento2#12342 : JSTestDriver removal

### Manual testing scenarios
<!--- Provide a set of unambiguous steps to test the proposed code change -->
1. Make sure that NPM is installed, if not `$ npm install`.
2. Generate static view files in Magento that are going to be tested `$ php bin/magento setup:static-content:deploy -f`.
3. Run Jasmine testing `$ grunt spec:`.

All the tests should pass.

### List of JS which require attention:

- [x] [JsTestDriver/testsuite/mage/accordion/accordion.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/accordion/accordion.js)
- [x] [JsTestDriver/testsuite/mage/collapsible/test-collapsible.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/collapsible/test-collapsible.js)
- [x] [JsTestDriver/testsuite/mage/dropdown/test-dropdown.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/dropdown/test-dropdown.js)
- [x] [JsTestDriver/testsuite/mage/form/form-test.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/form/form-test.js)
- [x] [JsTestDriver/testsuite/mage/menu/test-menu.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/menu/test-menu.js)
- [x] [JsTestDriver/testsuite/mage/translate/translate-test.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/translate/translate-test.js)
- [x] [JsTestDriver/testsuite/mage/translate_inline/translate-inline-test.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/translate_inline/translate-inline-test.js)
- [ ] [JsTestDriver/testsuite/mage/validation/test-validation.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/validation/test-validation.js)
- [x] [JsTestDriver/testsuite/mage/decorate-test.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/decorate-test.js)
- [x] [JsTestDriver/testsuite/mage/calendar](https://github.com/magento/magento2/tree/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/calendar)
- [x] [JsTestDriver/testsuite/mage/loader](https://github.com/magento/magento2/tree/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/loader)
- [x] [JsTestDriver/testsuite/mage/suggest](https://github.com/magento/magento2/tree/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/suggest)
- [x] [JsTestDriver/testsuite/mage/tabs](https://github.com/magento/magento2/tree/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/tabs)

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

Commit:
Added translate.test.js for Jasmine compatibility unit testing.
Removed transtale-test.js from jsTestDriver suite.

Commit:
Added translate.test.js for Jasmine compatibility unit testing.
Removed transtale-test.js from jsTestDriver suite.

Commit:
Fix translate.test to be eslint friendly

Commit:
Fix translate.test to be eslint friendly

KarlDeux commented:
@ishakhsuvarov are we still using [JsTestDriver/testsuite/mage/translate_inline/translate-inline-test.js](https://github.com/magento/magento2/blob/2.2-develop/dev/tests/js/JsTestDriver/testsuite/mage/translate_inline/translate-inline-test.js) ?

Couldn't find any usage into the magento 2 and those tests are hard to deal with since that'd involve DOM inclusions. Without jasmine-jquery it'd be hard.

Commit:
Added accordion.test.js for Jasmine testing.
Removed accordion.js and index.html for jsTestDriver test.

Commit:
Added accordion.test.js for Jasmine testing.
Removed accordion.js and index.html for jsTestDriver test.

Commit:
Eslint upgrades.

Commit:
Eslint upgrades.

omiroshnichenko commented:
Hi, @KarlDeux. Thank you for your interest in this task. We still using translate-inline in [admin](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Translation/view/adminhtml/templates/translate_inline.phtml#L56) and [front](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Translation/view/frontend/templates/translate_inline.phtml#L59) areas. For such case, you can create separate html file and fetch it using require text plugin.

KarlDeux commented:
Ok @omiroshnichenko will do so.

omiroshnichenko commented:
@KarlDeux Im have already rewritten decorate-test.js. Is it be ok if I will make commit to your branch?

KarlDeux commented:
@omiroshnichenko sure thing!

omiroshnichenko committed:
Added decorate.test.js for Jasmine testing. Removed decorate-test.js

omiroshnichenko committed:
Added decorate.test.js for Jasmine testing. Removed decorate-test.js

(cherry picked from commit b3fedbc)

Commit:
Eslint improvements and logical fixes for unit testing.

Commit:
Merge remote-tracking branch 'karldeux/issue/12342/JSTestDriver-removal' into issue/12342/JSTestDriver-removal

# Conflicts:
#	dev/tests/js/jasmine/tests/lib/mage/accordion.test.js

Commit:
Eslint improvements and logical fixes for unit testing.

KarlDeux commented:
@omiroshnichenko looks like decorate test are failing:

       Error: Method customMethod does not exist on jQuery.decorate in http://localhost:8000/pub/static/frontend/Magento/blank/en_US/jquery.js (line 253) (1)

omiroshnichenko committed:
Fix decorate.test.js

omiroshnichenko commented:
@KarlDeux My fault. Fixed.

Commit:
Fixed unsafe comparison with a literal. No need to eslint eqeqeq.

Fixed enable method. It was not enabling and activating the collapsible.
This was detected on Jasmine unit testing.

Minor repositioning to this.options.disabled on disable method for
consistency purposes with enable method.

Commit:
that - this consistency and splat complete function to be Jasmine friendly.

Commit:
Added collapsible.test.js and removed JsTestDriver equivalent.

KarlDeux commented:
Had to apply some changes to [lib/web/mage/collapsible.js](https://github.com/magento/magento2/blob/2.2-develop/lib/web/mage/collapsible.js) in order to be testable. 
Also discovered that enable function wasn't working properly running the tests. Did some changes so it is actually enabling the chollapsible item (enabling tab-index, showing the element), dunno if this is the actual behavior that should be applied but reading the previous test cases it'd make sense.

Commit:
jscs:disable jsDoc

KarlDeux commented:
Dropdown test in progress.

Commit:
Added dropdown.test.js and removed JsTestDriver equivalent.

Commit:
Backend jquery-ui version has a diferent behavior on create that will add ui-dialog-titlebar anyway.
This test is only valid with frontend's version
@TODO: Re-design Jasmine settings to match a selective architecture for testing in order to split them.

KarlDeux commented:
@omiroshnichenko looks like I've found a "stopper". 

One of the tests from dropdown.test.js was failing on backend test due to the differences with jquery-ui version. Frontend version won't add on create ui-dialog-titlebar while backend will do so.

I decided to apply a selective expectation based on jquery-ui version instead of re-designing the jasmine configuration for lib folder and applying structural changes to do so. (ie: lib/mage/frontend/**/*.test.js, lib/mage/backend/**/*.test.js ... )

What do you think?

`
// Backend jquery-ui version has a diferent behavior on create that will add ui-dialog-titlebar anyway.
        // This test is only valid with frontend's version
        // @TODO: Re-design Jasmine settings to match a selective architecture for testing in order to split them.
        if ($.ui.version !== '1.9.2') {
            it('check if the title bar is prevented from being created', function () {
                var dialog = $('<div/>'),
                    uiClass = '.ui-dialog',
                    ui;

                dialog.dropdownDialog({
                    'createTitleBar': true
                });

                ui = $(uiClass);
                expect(ui.find('.ui-dialog-titlebar').length > 0).toBeTruthy();

                dialog.dropdownDialog('destroy');

                dialog.dropdownDialog({
                    'createTitleBar': false
                });

                ui = $(uiClass);
                expect(ui.find('.ui-dialog-titlebar').length <= 0).toBeTruthy();

                dialog.dropdownDialog('destroy');
            });
        }
`

KarlDeux commented on dev/tests/js/jasmine/tests/lib/mage/dropdown.test.js:
```diff
@@ -270,29 +270,34 @@ define([
-            var dialog = $('<div/>'),
-                uiClass = '.ui-dialog',
-                ui;
+        // Backend jquery-ui version has a diferent behavior on create that will add ui-dialog-titlebar anyway.
```
@omiroshnichenko looks like I've found a "stopper". 

One of the tests from dropdown.test.js was failing on backend test due to the differences with jquery-ui version. Frontend version won't add on create ui-dialog-titlebar while backend will do so.

I decided to apply a selective expectation based on jquery-ui version instead of re-designing the jasmine configuration for lib folder and applying structural changes to do so. (ie: lib/mage/frontend/**/*.test.js, lib/mage/backend/**/*.test.js ... )

What do you think?

omiroshnichenko replied:
@KarlDeux Your solution looks good for me. Re-designing jasmine configuration for lib folder it is incorrect because lib folder can be used in both areas.

KarlDeux commented:
@omiroshnichenko shall we cover with test deprecated libraries such as edit-trigger ?
That one is deprecated since version 2.2.0.

omiroshnichenko commented:
@KarlDeux No, you can skip it.

Commit:
Added form.test.js and removed JsTestDriver equivalent.
Modified if documentation to be more friendly with developers.

Commit:
jQuery consistency.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added menu.test.js and removed JsTestDriver equivalent.

KarlDeux commented:
@omiroshnichenko i'm with the validation test... it's just huge haha

omiroshnichenko commented:
I will start on translate-inline.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added translate-inline.test.js and removed JsTestDriver equivalent.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added calendar.test.js and removed JsTestDriver equivalent.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Fix translate-inline.test.js, save original ajax

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Fix loader.test.js and removed JsTestDriver equivalent.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added suggest.test.js and removed JsTestDriver equivalent.
- Added tree-suggest.test.js and removed JsTestDriver equivalent.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added tabs.test.js and removed JsTestDriver equivalent.

omiroshnichenko commented:
@KarlDeux Maybe I can help you with validation test?

omiroshnichenko committed:
Merge branch '2.2-develop' of github.com:magento/magento2 into issue/12342/JSTestDriver-removal

omiroshnichenko commented:
@KarlDeux If you busy, I can start work on validation-test?

KarlDeux commented:
@omiroshnichenko hey! I was out for vacations, just commited the file as I have it (still needs work). Feel free to finish it :)

omiroshnichenko commented:
@KarlDeux Ohhh!!! I just finished all test and ready to commit. Maybe I will make force push if you do not mind?

KarlDeux commented:
Sure @omiroshnichenko ! go for it!

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Added validation.test.js and removed JsTestDriver equivalent.
- Fix unstable tests.

omiroshnichenko committed:
Merge branch '2.2-develop' of github.com:magento/magento2 into issue/12342/JSTestDriver-removal

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Fix timeout dropdown close.

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Fix jstree override.

omiroshnichenko committed:
Merge branch '2.2-develop' of github.com:magento/magento2 into issue/12342/JSTestDriver-removal

omiroshnichenko committed:
magento/magento2#12342: JSTestDriver removal
- Fix validation.

magento-engcom-team commented:
@KarlDeux thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

KarlDeux commented:
@omiroshnichenko guess we're done right? Shall I squash it all or do you want to do so?

ishakhsuvarov commented:
@KarlDeux Going to merge it as is, don't see any problem in keeping history for this amount of work done 👍 
Thank you!
